[Merged by Bors] - feat(ring_theory/simple_module): Simple modules as simple objects in the Module category

pbazin opened the pull request:
A simple module is a simple object in the Module category. 
From there it should be easy to write an alternative proof of Schur's lemma for modules using category theory (using the work already done in category_theory/preadditive/schur.lean).
The other direction (a simple object in the Module category is a simple module) hasn't been proved yet.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

pbazin committed:
Proof a simple module is a simple object in the Module category.

jcommelin commented on src/ring_theory/simple_module.lean:
```diff
@@ -91,6 +93,32 @@ theorem is_semisimple_iff_top_eq_Sup_simples :
+    { cases eq_bot_or_eq_top f.range,
+      { exfalso, apply h, rw ← linear_map.range_eq_bot, exact h_1 },
+      { rw epi_iff_range_eq_top, exact h_1 } },
+    { apply is_iso_of_mono_of_epi } } end}
```
```suggestion
    { apply is_iso_of_mono_of_epi } }
  end }
```

jcommelin commented on src/ring_theory/simple_module.lean:
```diff
@@ -91,6 +93,32 @@ theorem is_semisimple_iff_top_eq_Sup_simples :
+    haveI : is_simple_order (submodule R (of R M)) := _inst,
+    haveI : epi f,
+    { cases eq_bot_or_eq_top f.range,
+      { exfalso, apply h, rw ← linear_map.range_eq_bot, exact h_1 },
```
Please use `cases ... with h₁ h₂` on the line above, so that you explicitly give a name to the variables. On this line you use `h_1`, which is an auto-generated name and hence not so robust from a proof-maintenance point of view.

pbazin committed:
Rewriting proof

eric-wieser commented on src/ring_theory/simple_module.lean:
```diff
@@ -91,6 +93,38 @@ theorem is_semisimple_iff_top_eq_Sup_simples :
+      exact is_iso_of_mono_of_epi _ }
+  end }
+
+end category
```
I'd argue that this stuff belongs somewhere in the `algebra.category.Module` folder, rather than making this part of mathlib depend on category theory.

pbazin replied:
For some reason I'm not able/don't know how to create a new file in the PR so I wasn't able to move this to a new file in algebra/category/Module

eric-wieser replied:
How are you using `git` at the moment?

pbazin replied:
I'm using instructions at https://leanprover-community.github.io/contribute/index.html#making-a-pull-request-pr,  adding new commits by git commit -a + git push origin.

On VSCode it says "untracked file" for my new file if it helps

eric-wieser replied:
`git add yourfile.lean` before the commit step will do the trick

pbazin replied:
Thanks !

pbazin committed:
Moved changes to algebra/category/Module

pbazin committed:
It seems my new file doesn't appear in the last commit on github.
Cancelling the last commit.

pbazin committed:
Moved changes to algebra/category/Module

pbazin committed:
(Not tested locally as it requires recompiling mathlib
due to modification of core files)
Added lemmas on surjective constant functions.

eric-wieser committed:
Merge branch 'master' into simple_simple_module

pbazin committed:
This version should be working but should still be golfable.

pbazin committed:
Merge branch 'simple_simple_module' of https://github.com/leanprover-community/mathlib into simple_simple_module

Vierkantor commented on src/logic/function/basic.lean:
```diff
@@ -155,6 +155,10 @@ protected theorem surjective.exists₃ (hf : surjective f) {p : β → β → β
 
+lemma subsingleton_of_surjective_const (α : Type*) {β : Type*} (b : β)
+  (h : function.surjective (function.const α b)) : subsingleton β :=
+subsingleton_of_forall_eq b $ h.forall.mpr (λ _, rfl)
```
You can strengthen this to `unique`, since we already have `b : β`.

Vierkantor commented on src/algebra/category/Module/simple.lean:
```diff
@@ -0,0 +1,49 @@
+  apply subsingleton_of_surjective_zero N,
+  have : function.surjective (0 : N ⟶ of R M) := (Module.epi_iff_surjective _).mp ‹_›,
+  exact this,
+end
```
The `‹_›` need some help to find the right hypothesis, but just naming it will work:
```suggestion
lemma subsingleton_of_epi_zero (N : Module R) [h : epi (0 : N ⟶ of R M)] : subsingleton M :=
subsingleton_of_surjective_zero N ((Module.epi_iff_surjective _).mp h)
```

pbazin committed:
Golf subsingleton_of_epi_zero

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

pbazin committed:
Strengthened subsingleton to unique

pbazin committed:
Fix lint

pbazin committed:
Added docstrings

pbazin committed:
Added docstring

jcommelin commented on src/algebra/category/Module/simple.lean:
```diff
@@ -0,0 +1,46 @@
+
+/-- If the zero morphism is an epi then the codomain in trivial. -/
+def unique_of_epi_zero (N : Module R) [h : epi (0 : N ⟶ of R M)] : unique M :=
+unique_of_surjective_zero N ((Module.epi_iff_surjective _).mp h)
```
```suggestion
/-- If the zero morphism is an epi then the codomain is trivial. -/
def unique_of_epi_zero (N : Module R) [h : epi (0 : N ⟶ of R M)] : unique M :=
unique_of_surjective_zero N ((Module.epi_iff_surjective _).mp h)
```

jcommelin replied:
I think this should move to `Module/basic.lean`.

pbazin replied:
I moved  it to `Module/epi_mono.lean`, as that's where `epi_iff_surjective` is defined

jcommelin commented on src/algebra/category/Module/simple.lean:
```diff
@@ -0,0 +1,46 @@
+unique_of_surjective_zero N ((Module.epi_iff_surjective _).mp h)
+
+instance is_simple_module_of [_inst : is_simple_module R M] : is_simple_module R (of R M) :=
+_inst
```
```suggestion
instance is_simple_module_of [H : is_simple_module R M] : is_simple_module R (of R M) := H
```

pbazin committed:
Moved unique_of_epi_zero to epi_mono

pbazin committed:
bugfix

pbazin committed:
bugfix

jcommelin reviewed (approved):
Thanks :tada:

bors merge
